File: electron/tests/scenarios/snapshot.ts
import type { FakeLog } from './fakes'

/**
 * Dump the whole library state for a scenario snapshot. Everything comes out of `db.raw` so the
 * file shows exactly what is stored. Lists are ordered; absolute paths are rewritten to
 * `{library}/…`, `{corpus}/…`, `{fixture}` markers by the caller's `relativize` pass, and
 * remaining UUID-shaped strings become `uuid-N` in first-seen order (agent task runs call
 * `uuid()` directly — agent_runs.id, audit_log.agent_run_id, collection_items.agent_run_id and
 * relationships.agent_run_id are the fields that need it).
 */
export interface SnapshotContext {
  raw: { prepare(sql: string): { all(...args: unknown[]): unknown[] } }
  fakes: FakeLog
  settings: unknown
}

interface TableSpec {
  /** Output key. */
  key: string
  sql: string
}

const TABLES: TableSpec[] = [
  { key: 'items', sql: 'SELECT * FROM items ORDER BY id' },
  { key: 'fts', sql: 'SELECT * FROM items_fts ORDER BY item_id' },
  { key: 'collections', sql: 'SELECT * FROM collections ORDER BY id' },
  { key: 'collectionItems', sql: 'SELECT * FROM collection_items ORDER BY collection_id, item_id' },
  { key: 'relationships', sql: 'SELECT * FROM relationships ORDER BY id' },
  { key: 'suppressions', sql: 'SELECT * FROM suppressions ORDER BY kind, key' },
  { key: 'audit', sql: 'SELECT * FROM audit_log ORDER BY id' },
  // Insertion order, not id: run ids are real UUIDs, so ordering by id scrambles every run.
  { key: 'agentRuns', sql: 'SELECT * FROM agent_runs ORDER BY rowid' },
  { key: 'jobs', sql: 'SELECT * FROM jobs ORDER BY id' }
]

export function takeSnapshot(ctx: SnapshotContext): Record<string, unknown> {
  const state: Record<string, unknown> = {}
  for (const { key, sql } of TABLES) {
    state[key] = ctx.raw.prepare(sql).all()
  }
  state.settings = ctx.settings
  state.thumbnailCalls = ctx.fakes.thumbnailCalls
  state.snapshotCalls = ctx.fakes.snapshotCalls
  state.fetchCalls = ctx.fakes.fetchCalls
  state.desktopCalls = ctx.fakes.desktopCalls
  state.pushes = ctx.fakes.pushes
  return state
}

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/

/**
 * Determinism pass over the whole expected document: absolute path prefixes become markers and
 * UUID-shaped strings become `uuid-N` (shared mapping, so the same run id stays consistent
 * across agent_runs / audit_log / collection_items).
 */
export function normalizeDocument(doc: unknown, prefixes: { find: string; marker: string }[]): unknown {
  const uuids = new Map<string, number>()
  const map = new Map(prefixes.map((p) => [p.find, p.marker]))

  const visit = (value: unknown): unknown => {
    if (typeof value === 'string') {
      if (UUID_RE.test(value)) {
        if (!uuids.has(value)) uuids.set(value, uuids.size + 1)
        return `uuid-${uuids.get(value)}`
      }
      let v = value
      // Longest prefix first so nested roots (userData inside tmpdir) win.
      for (const [find, marker] of [...map.entries()].sort((a, b) => b[0].length - a[0].length)) {
        if (v === find) v = marker
        else if (v.startsWith(`${find}/`)) v = `${marker}/${v.slice(find.length + 1)}`
        else if (v.includes(`${find}/`)) v = v.split(`${find}/`).join(`${marker}/`)
      }
      // mkdtemp suffixes are random per run; timings inside JSON columns are wall-clock.
      return v
        .replace(/ka-(scenario|keep)-[\w-]+/g, (m) => `${m.replace(/-[A-Za-z0-9]+$/, '')}-X`)
        .replace(/\\?"?(durationMs|latencyMs)\\?":\s*\d+/g, '"$1":0')
    }
    if (Array.isArray(value)) return value.map(visit)
    if (value !== null && typeof value === 'object') {
      const out: Record<string, unknown> = {}
      for (const [k, v] of Object.entries(value)) {
        // `pid` in lock-owner records is a real process id; timings are wall-clock.
        out[k] = (k === 'pid' || k === 'durationMs' || k === 'latencyMs') && typeof v === 'number' ? 0 : visit(v)
      }
      return out
    }
    return value
  }
  return visit(doc)
}

File: Sources/KATestSupport/Snapshot.swift
import Foundation
import GRDB
import KAModel
import KAStorage

/// Port of `electron/tests/scenarios/snapshot.ts`: the 9 tables dumped verbatim, plus the
/// determinism pass (`normalize`) the runner applies before comparing to the expected file.
/// `settings`/fake-call keys are not part of this snapshot — those are runner/E2E state.
public enum Snapshot {
  private static let tables: [(key: String, sql: String)] = [
    ("items", "SELECT * FROM items ORDER BY id"),
    ("fts", "SELECT * FROM items_fts ORDER BY item_id"),
    ("collections", "SELECT * FROM collections ORDER BY id"),
    ("collectionItems", "SELECT * FROM collection_items ORDER BY collection_id, item_id"),
    ("relationships", "SELECT * FROM relationships ORDER BY id"),
    ("suppressions", "SELECT * FROM suppressions ORDER BY kind, key"),
    ("audit", "SELECT * FROM audit_log ORDER BY id"),
    // Insertion order, not id: run ids are real UUIDs, so ordering by id scrambles every run.
    ("agentRuns", "SELECT * FROM agent_runs ORDER BY rowid"),
    ("jobs", "SELECT * FROM jobs ORDER BY id")
  ]

  /// Table keys in snapshot order, for callers that compare against the expected file.
  public static var tableKeys: [String] { tables.map(\.key) }

  /// node:sqlite returns INTEGER/REAL -> number, TEXT -> string, NULL -> null, BLOB -> Uint8Array
  /// (JSON.stringify emits `{ "0": byte, ... }`, which is what the expected file would contain).
  private static func jsonValue(_ v: DatabaseValue) -> JSONValue {
    switch v.storage {
    case .null: return .null
    case .int64(let i): return .number(Double(i))
    case .double(let d): return .number(d)
    case .string(let s): return .string(s)
    case .blob(let data):
      var o = [String: JSONValue]()
      for (i, b) in data.enumerated() { o["\(i)"] = .number(Double(b)) }
      return .object(o)
    }
  }

  private static func rowToJson(_ row: Row) -> JSONValue {
    var o = [String: JSONValue]()
    for (name, value) in zip(row.columnNames, row.databaseValues) {
      o[name] = jsonValue(value)
    }
    return .object(o)
  }

  public static func take(_ db: Db) throws -> JSONValue {
    try db.readOnly { d in
      var state = [String: JSONValue]()
      for t in tables {
        state[t.key] = try .array(Row.fetchAll(d, sql: t.sql).map(rowToJson))
      }
      return .object(state)
    }
  }

  private static let uuidRegex = try! NSRegularExpression(
    pattern: "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
  private static let tempSuffixRegex = try! NSRegularExpression(pattern: "ka-(scenario|keep)-[\\w-]+")
  private static let trailingSegmentRegex = try! NSRegularExpression(pattern: "-[A-Za-z0-9]+$")
  private static let timingRegex = try! NSRegularExpression(
    pattern: "\\\\?\"?(durationMs|latencyMs)\\\\?\":\\s*\\d+")

  /// Same rules as `normalizeDocument`: UUIDs -> uuid-N first-seen, path prefixes -> markers
  /// (longest first, including mid-string occurrences), mkdtemp suffixes -> -X, timings -> 0,
  /// and numeric pid/durationMs/latencyMs fields -> 0.
  public static func normalize(_ doc: JSONValue, prefixes: [(find: String, marker: String)]) -> JSONValue {
    var uuids = [String: Int]()
    let sorted = prefixes.sorted { $0.find.count > $1.find.count }

    func normalizeString(_ s: String) -> String {
      var v = s
      if uuidRegex.firstMatch(in: v, range: NSRange(v.startIndex..., in: v)) != nil {
        if uuids[v] == nil { uuids[v] = uuids.count + 1 }
        return "uuid-\(uuids[v]!)"
      }
      for p in sorted {
        if v == p.find {
          v = p.marker
        } else if v.hasPrefix(p.find + "/") {
          v = p.marker + "/" + v.dropFirst(p.find.count + 1)
        } else if v.contains(p.find + "/") {
          v = v.replacingOccurrences(of: p.find + "/", with: p.marker + "/")
        }
      }
      v = tempSuffixRegex.replacingMatches(in: v) { whole, _ in
        let stripped = trailingSegmentRegex.stringByReplacingMatches(
          in: whole, range: NSRange(whole.startIndex..., in: whole), withTemplate: "")
        return stripped + "-X"
      }
      v = timingRegex.replacingMatches(in: v) { _, group in "\"\(group ?? "")\":0" }
      return v
    }

    func visit(_ value: JSONValue) -> JSONValue {
      switch value {
      case .string(let s): return .string(normalizeString(s))
      case .array(let a): return .array(a.map(visit))
      case .object(let o):
        var out = [String: JSONValue]()
        for (k, v) in o {
          if (k == "pid" || k == "durationMs" || k == "latencyMs"), case .number = v {
            out[k] = .number(0)
          } else {
            out[k] = visit(v)
          }
        }
        return .object(out)
      default: return value
      }
    }
    return visit(doc)
  }
}

private extension NSRegularExpression {
  /// Replace each match; `body` gets the whole match and first capture group (if any).
  func replacingMatches(in s: String, with body: (_ whole: String, _ group1: String?) -> String) -> String {
    let matches = matches(in: s, range: NSRange(s.startIndex..., in: s))
    var out = s
    for m in matches.reversed() {
      guard let whole = Range(m.range, in: out) else { continue }
      let group = m.numberOfRanges > 1 ? Range(m.range(at: 1), in: out).map { String(out[$0]) } : nil
      out.replaceSubrange(whole, with: body(String(out[whole]), group))
    }
    return out
  }
}
